### PR TITLE
Batch normalization class and widget

### PR DESCRIPTION
**Pull request by mstrazar**
##### Issue
Implements batch normalization, issue #123 .

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by mstrazar** [comment]
Since the batch effect correction is merely linear, this results in a single constant correction for each class value. This can have adverse effects on some pathological cases, where the expression distributions (conditioned on a class) are not normal. This is often the case in single cell data.

File: orangecontrib/single_cell/preprocess/scbnorm.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            assert all((a.name in self.models for a in atts))
+            assert (data.domain.index(b) for b in self.batch_vars)
+            Z = self._design_matrix(data)
+            W = np.hstack((self.models[a.name].reshape((Z.shape[1], 1)) for a in atts))
+            Xc = data.X.copy()
+            if self.nonzero_only:
+                nz = np.where(Xc)
+                Xn = INV_LINKS[self.link](LINKS[self.link](Xc[nz]) - Z.dot(W)[nz])
+                Xc[nz] = Xn
+
```
Comment (lanzagar) [inline orangecontrib/single_cell/preprocess/scbnorm.py]: empty line not needed

File: orangecontrib/single_cell/preprocess/scbnorm.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            else:
+                Xc = INV_LINKS[self.link](LINKS[self.link](Xc) - Z.dot(W))
+            return Table.from_numpy(domain=data.domain,
```
Comment (lanzagar) [inline orangecontrib/single_cell/preprocess/scbnorm.py]: `Table.from_numpy` should be avoided when possible for data transformations (instance ids are lost).
`data.transform` can be used in most circumstances, but here the domain actually stays the same so you can even use `data.copy()`. And then just overwrite the values in X.

File: orangecontrib/single_cell/widgets/ownormalization.py [hunk 3]
```diff
@@ -82,13 +105,18 @@ def set_data(self, data):
...
         self.normalize_check.setEnabled(len(self.attrs_model) > 0)
         self.combo_attrs.setEnabled(self.normalize_cells)
 
         self.Outputs.data.send(None)
         self.openContext(self.data.domain)
         self.on_changed()
 
+    def _batch_changed(self):
+        self.batch_vars = tuple(self.batch_attrs[ind.row()].name
+                                for ind in self.varview.selectionModel().selectedRows())
```
Comment (lanzagar) [inline orangecontrib/single_cell/widgets/ownormalization.py]: Changing batch normalization options (link or var) does not trigger a commit even when auto-commit is on.
Probably missing a `self.commit()` at the end here?

**Comment by lanzagar** [comment]
Normalize now crashes on dataset: Cell cycle in mouse liver (cell cycle genes)
(it does not on master)

**Pull request merged**